deepdrivemd/agents/stream/dbscan.py

```diff
@@ -0,0 +1,692 @@
+    velocities = cvae_input[3][outlier_list]
+    md5s = cvae_input[2][outlier_list]
+    rmsds = cvae_input[4][outlier_list]
+
+    z = list(zip(positions, velocities, md5s, rmsds, outlier_list))
+    indices = np.arange(len(z))
+    np.random.shuffle(indices)
+    indices = indices[:N]
+    z = [z[i] for i in indices]
+    z = list(zip(*z))
+
+    return z
+
+
+def select_best_random(
+    cfg: OutlierDetectionConfig,
+    cvae_input: Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray],
+) -> List[int]:
+    """Sort cvae_input by rmsd, selects :obj:`2*cfg.num_sim` best entries, out of them
+    randomly select :obj:`cfg.num_sim`, return the corresponding indices.
+
+    Parameters
+    ----------
+    cfg : OutlierDetectionConfig
+    cvae_input : Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]
+        steps, positions, velocities, md5sums, rmsds
+
+    Returns
+    -------
+    List[int]
+        List of :obj:`cfg.num_sim` indices randomly selected from a list of
+        :obj:`2*cfg.num_sim` entries with smallest rmsd.
+
+    Note
+    ----
+    This is used when no outliers are found.
+    """
+    rmsds = cvae_input[4]
+    z = sorted(zip(rmsds, range(len(rmsds))), key=lambda x: x[0])
+    sorted_index = list(map(lambda x: x[1], z))[2 * cfg.num_sim :]
+    sorted_index = random.sample(sorted_index, cfg.num_sim)
+    return sorted_index
+
+
+def select_best(
+    cfg: OutlierDetectionConfig,
+    cvae_input: Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray],
+) -> List[int]:
+    """Sort cvae_input by rmsd, selects best :obj:`cfg.num_sim`, return
+    the corresponding indices.
+
+    Parameters
+    ----------
+    cfg : OutlierDetectionConfig
+    cvae_input : Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]
+        steps, positions, velocities, md5sums, rmsds
+
+    Returns
+    -------
+    List[int]
+        List of :obj:`cfg.num_sim` indices for best traversed states among
+        :obj:`lastN` from each aggregator.
+    """
+    rmsds = cvae_input[4]
+    z = sorted(zip(rmsds, range(len(rmsds))), key=lambda x: x[0])
+    sorted_index = list(map(lambda x: x[1], z))[cfg.num_sim :]
+    return sorted_index
+
+
+def main(cfg: OutlierDetectionConfig):
+    print(subprocess.getstatusoutput("hostname")[1])
+    sys.stdout.flush()
+
+    print(cfg)
+    with Timer("wait_for_input"):
+        adios_files_list = wait_for_input(cfg)
+    with Timer("wait_for_model"):
+        model_path = str(wait_for_model(cfg))
+
+    variable_list = [
+        StreamContactMapVariable("contact_map", np.uint8, 1),
+        StreamVariable("positions", np.float32, 1),
+        StreamVariable("md5", str, 2),
+        StreamVariable("velocities", np.float32, 1),
+        StreamScalarVariable("rmsd", np.float32, 0),
+    ]
+
+    mystreams = Streams(
+        adios_files_list,
+        variable_list,
+        lastN=cfg.lastN,
+        config=cfg.adios_xml_agg,
```

Comment:
Same type issue as other comments. `cfg.adios_xml_agg` is Path but config expects str.

Reply:
That was changed in Streams.